### PR TITLE
Remove legacy download stats.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

File: internal/charmstore/migrations_integration_test.go [hunk 1]
```diff
@@ -11,10 +11,10 @@ import (
 	"gopkg.in/juju/charm.v6-unstable"
 	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
 	"gopkg.in/mgo.v2"
-	"gopkg.in/mgo.v2/bson"
 
 	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/storetesting"
+	"gopkg.in/mgo.v2/bson"
```
Comment (mhilton) [inline internal/charmstore/migrations_integration_test.go]: This shouldn't have moved.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1784/
Test PASSed.

**Comment by mhilton** [comment]
:+1:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1794/
Test PASSed.

File: internal/charmstore/migrations_integration_test.go (hunk 1 above)
Comment (fabricematrat) [inline internal/charmstore/migrations_integration_test.go]: Done.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1795/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1796/
Test PASSed.

File: internal/charmstore/migrations.go [hunk 2]
```diff
@@ -350,6 +357,77 @@ func ncmUpdateBaseEntity(db StoreDatabase, baseEntity *preNCMBaseEntity) error {
...
+			}
+		}
+	}
+	if err := iter.Err(); err != nil {
+		return errgo.Notef(err, "cannot iterate through entities")
+	}
+	return nil
+}
+
+func setArchiveDownloadStats(db StoreDatabase, key string, legacyDownloadStats []byte) error {
```
Comment (jrwren) [inline internal/charmstore/migrations.go]: Does this set, or inc? if it increments, name it inc instead of set?

**Comment by jrwren** [comment]
👍

File: internal/charmstore/migrations.go (hunk 2 above)
Comment (fabricematrat) [inline internal/charmstore/migrations.go]: @jrwren  In fact it set a number of download for a specific date/hours/minute.
Then the meta/stats aggregate the data so set seems correct

Comment (rogpeppe) [inline internal/charmstore/migrations.go]: Well, technically, I think @jrwren is correct - if there were already any downloads
for that specific date, this will add to them not set them, right?

File: internal/charmstore/migrations_integration_test.go [hunk 3]
```diff
@@ -386,6 +450,26 @@ var migrationFromDumpBaseEntityTests = []struct {
 	},
 }}
 
+var migrationFromDumpStatsTests = []struct {
+	id    string
+	total int64
+}{{
```
Comment (rogpeppe) [inline internal/charmstore/migrations_integration_test.go:456]: Perhaps it would be nicer if these all had different totals, so we could tell that the right totals were being assigned to their respective entities?

File: internal/charmstore/stats.go [hunk 4]
```diff
@@ -528,40 +513,6 @@ func (s *Store) statsCacheFetch(id *charm.URL) (interface{}, error) {
 	if err != nil {
 		return nil, errgo.Notef(err, "cannot get aggregated count for %q", id)
 	}
-	if !LegacyDownloadCountsEnabled {
-		return counts, nil
-	}
-	// TODO (frankban): remove this code when removing the legacy counts logic.
```
Comment (rogpeppe) [inline internal/charmstore/stats.go:534]: Oooh it's good to see this go!

File: internal/v5/api.go [hunk 5]
```diff
@@ -515,10 +515,6 @@ func (h *ReqHandler) updateEntity(id *router.ResolvedURL, fields map[string]inte
 	if err != nil {
 		return errgo.Notef(err, "cannot update %q", &id.URL)
 	}
-	err = h.Store.UpdateSearchFields(id, fields)
```
Comment (rogpeppe) [inline internal/v5/api.go:518]: Are you sure it's appropriate to remove this?

Comment (rogpeppe) [inline internal/v5/api.go:518]: Are you sure it's appropriate to remove this?

Comment (fabricematrat) [inline internal/v5/api.go:518]: yes the only field checked was legacy-download-stats

File: internal/charmstore/migrations.go (hunk 2 above)
Comment (fabricematrat) [inline internal/charmstore/migrations.go]: yes after speaking I agree it's an increment.

Comment (fabricematrat) [inline internal/charmstore/migrations.go]: Renamed

File: internal/charmstore/migrations_integration_test.go (hunk 3 above)
Comment (fabricematrat) [inline internal/charmstore/migrations_integration_test.go:456]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1797/
Test PASSed.

**Comment by fabricematrat** [comment]
QA Instructions.

Have a database with lots of data 
You could for example take a dump of our QA database without the blog, dump of entities, base_entities, juju.stat.counters juju.stat.tokens collection.
Run the charmstore before the migration on v5-unstable branch for example.
Select a set of charm which have the extra-info legacy-download-stats set, some on latest revsion some on previous revision, charms without this extra-info but with some download-stats. Check the check the meta/stats.

Restart charmstore with the branch and check the meta/stats again to see if this is now corrected meaning you get the data only from download-stats, you could even remove the extra-info.legacy-download-stats from the revision to check you get always the same data. Check edge case like bundle, promulgated charm, non promulgated charm.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1798/
Test PASSed.

**Comment by frankban** [comment]
QA ok now. Stats seem sane even after removing legacy-download-stats from the charm extra-infos.

**Comment by fabricematrat** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge
